WIP Add OTE for origin tests

camilamacedo86 opened the pull request.

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+
+func main() {
+	registry := e.NewRegistry()
+	ext := e.NewExtension("operator-framework", "payload", "olmv1-tests-extension")
```
@jianzhangbjz @kuiwang

Could you please lend a hand on this one?
Are those the names?
What names should we add?

kuiwang02 commented:
@camilamacedo86 thanks. I will review it.

kuiwang02 commented:
after quickly going through this PR, I need to take more time review it. thanks

kuiwang02 commented:
I will finish review tomorrow because I need to take more time review it.
please keep the PR unchanged, thanks

camilamacedo86 commented:
Hi @kuiwang02

> @camilamacedo86 thanks. I will review it.
>I will finish review tomorrow because I need to take more time review it.
please keep the PR unchanged, thanks

Thank you for checking the PR! 😊

I needed to change it because now we can use Go 1.24. So, I rebased it and also renamed the bin to match the others — like in this example: https://github.com/openshift/origin/pull/29949/files.

Please note: this PR only adds the bin. There are no tests.
To try it, just pull the PR and run make build — that will give you the bin locally.
(You can also check the PR description.)

I didn’t understand the concern, since we are not using OCP, mocks, or special test cases here.
I just wanted to let you know, so you can fetch it again if needed.

Thanks a lot for your help and understanding! 🙏

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+
+func main() {
+	registry := e.NewRegistry()
+	ext := e.NewExtension("olm", "payload", "olmv1-tests-extension")
```
I found an example: https://github.com/openshift/machine-api-operator/blob/c0c93644a44a9a30cfc438f4a5cf6cb66b700421/cmd/machine-api-tests-ext/main.go#L35C38-L35C86

So, I updated it to opnshift, payload, operator-controller

But I am not sure if the component should be operator-controller, olmv1 or olm.
It might be something that if you do not know it, I might need to find out and TBD the value.

anik120 reviewed (changes requested):
Do we need to check in the `tests/extension/vendor` folder? I understand that we are checking in the top level vendor folder for this repo, but that's probably because of some downstream-ing process quirks.  

It'll be cleaner to have the vendor folder git ignored, and then the Make recipes to include `go mod vendor` before building the binary, if possible

camilamacedo86 commented:
HI @anik120 

> Do we need to check in the tests/extension/vendor folder? I understand that we are checking in the top level vendor folder for this repo, but that's probably because of some downstream-ing process quirks.

> It'll be cleaner to have the vendor folder git ignored, and then the Make recipes to include go mod vendor before building the binary, if possible

I agree with you 100%
But things does not work that is the same case of : https://github.com/openshift/operator-framework-operator-controller/pull/391 (default catalog tests)
I really would like to remove the vendor but seems that OCP ci things will ensure that we have that.

camilamacedo86 commented:
Hi @anik120 

The problem is that we call `make verify` from the root. Indeed, if we try to change the dir like https://github.com/openshift/operator-framework-operator-controller/pull/391 it will always fail in https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_operator-framework-operator-controller/391/pull-ci-openshift-operator-framework-operator-controller-main-verify/1940160259024752640 

Because of that we need to have the vendor :-( 
c/c @tmshort

kuiwang02 commented on openshift/tests-extension/Makefile:
```diff
@@ -0,0 +1,69 @@
+	# It prevents various FIPS compliance policies from being applied to this compilation.
+	# Do not set globally.
+	@mkdir -p $(TOOLS_BIN_DIR)
+	GO_COMPLIANCE_POLICY="exempt_all" go build -ldflags "$(LDFLAGS)" -o $(TOOLS_BIN_DIR)/operator-controller-tests-ext ./cmd/...
```
I suggest to change `operator-controller-tests-ext` to `olmv1-tests-ext`.
I know you name it as operator-controller-tests-ext because the repo is operator-controller.

but actually this repo does not only include the code of operator-controller, but also the code of catalogd.
so I think olmv1-tests-ext is better.
it avoid the misunderstanding: when someone see operator-controller-tests-ext, (s)he think it is only testing code of o-c.

camilamacedo86 replied:
I am fine with we change the name done 👍

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+		Parents: []string{"openshift/conformance/parallel"},
+	})
+
+	specs, err := g.BuildExtensionTestSpecsFromOpenShiftGinkgoSuite()
```
IHMO lets add it when we have the scenarios that will make sense to have it 
But it is great to know, really, thank you for sharing it !!!

kuiwang02 replied:
that's OK. thanks

camilamacedo86 replied:
I end up adding with a comment to clarify WHY it is required
Please, let me know if that makes sense
Marking as sorted out, too.

camilamacedo86 commented on openshift/tests-extension/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+
+FROM registry.ci.openshift.org/ocp/4.20:tools
+COPY --from=builder /tmp/build/operator-controller-tests-ext.gz /usr/bin/operator-controller-tests-ext.gz
+RUN gunzip /usr/bin/operator-controller-tests-ext.gz && chmod +x /usr/bin/operator-controller-tests-ext.gz
```
good catcher solved 
So I am closing this one

kuiwang02 replied:
ok, thanks

camilamacedo86 commented on openshift/tests-extension/Makefile:
```diff
@@ -0,0 +1,69 @@
+
+# Definitions for the extended tests
+
+GO_PKG_NAME := github/operator-framework-operator-controller/openshift/extended-tests
```
OH, I see, thanks solved.
Closing this one 👍

kuiwang02 replied:
ok, thanks

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+		panic(fmt.Sprintf("couldn't build extension test specs from ginkgo: %+v", err.Error()))
+	}
+
+	ext.AddSpecs(specs)
```
> You do not init testing framework, so you can not run the case with external binary if the case access cluster.

I'm not sure I fully understand what you meant there. From my understanding, if we have a running local cluster, we should be able to call the bin and observe the tests being executed.

That said, I feel we might be going a bit beyond the scope of this PR. The current goal is just to add the bin, and ensure it works correctly and allows us to validate the integration.

Any special or edge cases can be explored and tested more thoroughly when we implement that specific logic in the future. Otherwise, things could start to get confusing and harder to follow — kind of like putting the cart before the horse 😊

See that is how it is suggested in the documentation
So, I will keep it as it is until we have a valid reason to change it.
I hope that make sense

camilamacedo86 replied:
I tried to capture in the code with a TODO 👍

camilamacedo86 commented on openshift/tests-extension/test/olmv1.go:
```diff
@@ -0,0 +1,16 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("[sig-operator] OLMv1", func() {
```
It is just for us to test out.
But it is cool, I changed it, so marking it as sorted out.

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+For further information, please refer to the documentation at:
+https://github.com/openshift-eng/openshift-tests-extension/blob/main/cmd/example-tests/main.go
+*/
+package main
```
Thank you I added a note to capture the info shared 👍

camilamacedo86 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+	// This suite can be executed by running the following command:
+	// - $ make build
+	// - $ ./bin/tests-extension run-suite olmv1/tests
+	ext.AddSuite(e.Suite{
```
See that I updated to capture all info provided and we have a suite to cover all cases scenarios so I am marking this one as sorted out. Thank you.

kuiwang02 replied:
ok, thanks

camilamacedo86 commented on openshift/tests-extension/Makefile:
```diff
@@ -0,0 +1,69 @@
+	# It prevents various FIPS compliance policies from being applied to this compilation.
+	# Do not set globally.
+	@mkdir -p $(TOOLS_BIN_DIR)
+	GO_COMPLIANCE_POLICY="exempt_all" go build -ldflags "$(LDFLAGS)" -o $(TOOLS_BIN_DIR)/operator-controller-tests-ext ./cmd/...
```
Hi @kuiwang02 👋

Thanks for the info on `update` and why it's important for TestID tracking.

That makes sense, but when I run it locally, it generates metadata with absolute paths from my machine like:

```
[
  {
    "name": "[sig-olmv1] OLMv1 should pass a trivial sanity check",
    "labels": {},
    "resources": {
      "isolation": {}
    },
    "source": "openshift:payload:olmv1",
    "codeLocations": [
      "/Users/camilam/go/src/github/operator-framework-operator-controller/openshift/tests-extension/test/olmv1.go:12",
      "/Users/camilam/go/src/github/operator-framework-operator-controller/openshift/tests-extension/test/olmv1.go:13"
    ],
    "lifecycle": "blocking",
    "environmentSelector": {}
  }
]
```

This doesn’t work well for local dev because we often need to:

a) Build and test locally  
b) Rename or change specs  
c) Iterate quickly without metadata issues

If `update` runs inside `make build`, it breaks that flow — even small changes fail with “missing test” errors.

To fix that, I’ll open a JIRA so we can define a proper process for metadata and traceability.  
For now, I’ll avoid running `update` automatically in the build step.

We’ll need further discussion to align on a solution that balances compliance with a smooth dev experience.

Also — where should this metadata be generated and stored?  
It definitely shouldn’t live only in local paths, so that’s another piece we’ll need to clarify.

camilamacedo86 replied:
https://issues.redhat.com/browse/OPRUN-3992

camilamacedo86 replied:
Raised as well: https://redhat-internal.slack.com/archives/C07RDCVEYJG/p1751637388362499

kuiwang02 replied:
Hi @camilamacedo86 
I know you are raising the issue in slack, 
so, here is my suggestion now.
we do not add `update` into build, and make new cmd for update the combined. like
```console
.PHONY: build
build: build-binary update
.PHONY: build-binary
build-binary: #HELP Build the extended tests binary
	@mkdir -p $(TOOLS_BIN_DIR)
	GO_COMPLIANCE_POLICY="exempt_all" go build -ldflags "$(LDFLAGS)" -mod=vendor -o $(TOOLS_BIN_DIR)/olmv1-tests-ext ./cmd/...
.PHONY: update
update: 
	$(TOOLS_BIN_DIR)/olmv1-tests-ext update
```
it will not break any prow job because the json does not exist for every prow to make build.
actually OTE's purpose is to ensure test id unique with local json when we develop case, so we'd better to run update for each build locally. 

so, I suggest build, build-binary and update. and then  the member could use `build` to make case locally to ensure it is correct. (if the member does not want to use it, could use `build-binary`. it is not suggested)

so, 
1, it will not break any prow job because of test's name change
2, test id need to be checked locally during developing case. so, need case owner to run update after each building binary.

camilamacedo86 replied:
Hi @kuiwang02

Yes, we should not add in the build,
by ading we need to have coeshion and have something like

```
.PHONY: update-metadata
update-metadata: #HELP Build and run 'update-metadata' to generate test metadata
	$(TOOLS_BIN_DIR)/olmv1-tests-ext update --component openshift:payload:olmv1

.PHONY: build-update
build-update: #HELP Build the test binary and update metadata (used in CI or after test changes)
	$(MAKE) build
	$(MAKE) update-metadata
``` 

Anyway, it does not sort all concerns at all.

The Unique ID will use the path, and currently, **it generates the full path rather than the relative path.** This means that if I run this command on my machine, the result will not be the same if you run it on yours.** Therefore, it will not work at all, for example.

kuiwang02 commented on openshift/tests-extension/Makefile:
```diff
@@ -0,0 +1,69 @@
+	# It prevents various FIPS compliance policies from being applied to this compilation.
+	# Do not set globally.
+	@mkdir -p $(TOOLS_BIN_DIR)
+	GO_COMPLIANCE_POLICY="exempt_all" go build -ldflags "$(LDFLAGS)" -o $(TOOLS_BIN_DIR)/operator-controller-tests-ext ./cmd/...
```
ok, thanks

kuiwang02 commented on openshift/tests-extension/cmd/main.go:
```diff
@@ -0,0 +1,57 @@
+
+func main() {
+	registry := e.NewRegistry()
+	ext := e.NewExtension("openshift", "payload", "operator-controller")
```
ok, thanks

anik120 reviewed (commented):
@camilamacedo86 makes sense, thanks for checking that and confirming. 

/lgtm

kuiwang02 commented on openshift/tests-extension/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+FROM registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.24-openshift-4.20 AS builder
```
@camilamacedo86 reminder ^, thanks

kuiwang02 replied:
@camilamacedo86 
by the way, the https://github.com/openshift/operator-framework-operator-controller/blob/main/openshift/tests-extension/Dockerfile is not used for openshift-tests as I said [before](https://github.com/openshift/operator-framework-operator-controller/pull/383#discussion_r2179104727) `2, could keep this Dockerfile so that you could get the image of external image for your prow job.`. if you want to keep it, it only is used for your own prow testing if you would like. or else, you could remove it.

camilamacedo86 replied:
Hi @kuiwang

Thank you a lot. About the comments:

We do not need the dockerfile to run the tests on the prow it is working without this complexity, see: https://github.com/openshift/release/blob/master/ci-operator/config/openshift-priv/operator-framework-operator-controller/openshift-priv-operator-framework-operator-controller-main.yaml#L168-L181 AND https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_operator-framework-operator-controller/401/pull-ci-openshift-operator-framework-operator-controller-main-tests-extension/1943271877275291648

So, what you are saying is that we have no way of just adding the test bin to the Dockerfile, we must add the test ext bin in the Dockerfile of operator-controller. I would like to keep it isolated. However, so far, I have only found examples that do what you say as well.
I did : https://github.com/openshift/operator-framework-operator-controller/pull/403 now

Also, I am trying /testwith openshift/operator-framework-operator-controlle#403  in https://github.com/openshift/origin/pull/29949 lets see if that works.

kuiwang02 replied:
@camilamacedo86 

> However, so far, I have only found examples that do what you say as well.

Frankly I provided you the example of machine-api-operator in the [comment](https://github.com/openshift/operator-framework-operator-controller/pull/383#discussion_r2179104727) first time:

` suppose we choose the image olm-operator-controller, so we need to modify the [Dockerfile](https://github.com/openshift/operator-framework-operator-controller/blob/main/openshift/operator-controller.Dockerfile), like what this [Dockerfile](https://github.com/openshift/machine-api-operator/blob/main/Dockerfile.rhel) did.`

maybe you miss this in the comment.

here are exact example machine-api-operator:
https://github.com/openshift/machine-api-operator/blob/main/Dockerfile.rhel#L5-L7 to build machine-api-tests-ext
https://github.com/openshift/machine-api-operator/blob/main/Dockerfile.rhel#L16 to put it in component image.

by the way, why machine-api-operator do it in this way?
machine-api-operator team follow https://github.com/openshift/enhancements/pull/1676's requirment to add their binary into their component image. here we do same thing.

kuiwang02 commented:
> @camilamacedo86 makes sense, thanks for checking that and confirming.
> 
> /lgtm

reminder:
before merging PR, need to resolve https://github.com/openshift/operator-framework-operator-controller/pull/383#discussion_r2179104727.

perdasilva commented:
/approve
/lgtm
